builtin/credential/aws/path_role.go

```diff
@@ -218,82 +220,81 @@ func pathListRoles(b *backend) *framework.Path {
-		// written to between releasing the read lock and acquiring the write lock
-		roleEntry, err = b.nonLockedAWSRole(ctx, s, roleName)
-		if err != nil {
-			return nil, err
+	roleEntryRaw, found := b.roleCache.Get(roleName)
+	if found && roleEntryRaw != nil {
+		roleEntry, ok := roleEntryRaw.(*awsRoleEntry)
+		if !ok {
+			return nil, errors.New("could not convert role entry internally")
 		}
-		// somebody deleted the role, so no use in putting it back
 		if roleEntry == nil {
-			return nil, nil
+			return nil, errors.New("converted role entry is nil")
 		}
-		// now re-check to see if we need to upgrade
-		if needUpgrade, err = b.upgradeRoleEntry(ctx, s, roleEntry); err != nil {
-			return nil, errwrap.Wrapf("error upgrading roleEntry: {{err}}", err)
+	}
+
+	// Not found, or was nil
+	b.roleMutex.Lock()
+	defer b.roleMutex.Unlock()
+
+	return b.roleInternal(ctx, s, roleName)
+}
+
+// roleInternal does not perform locking, and rechecks the cache, going to disk if necessar
+func (b *backend) roleInternal(ctx context.Context, s logical.Storage, roleName string) (*awsRoleEntry, error) {
+	// Check cache again now that we have the lock
+	roleEntryRaw, found := b.roleCache.Get(roleName)
+	if found && roleEntryRaw != nil {
+		roleEntry, ok := roleEntryRaw.(*awsRoleEntry)
+		if !ok {
+			return nil, errors.New("could not convert role entry internally")
 		}
-		if needUpgrade {
-			if err = b.nonLockedSetAWSRole(ctx, s, roleName, roleEntry); err != nil {
-				return nil, errwrap.Wrapf("error saving upgraded roleEntry: {{err}}", err)
-			}
+		if roleEntry == nil {
+			return nil, errors.New("converted role entry is nil")
 		}
 	}
-	return roleEntry, nil
-}
 
-// lockedSetAWSRole creates or updates a role in the storage. This method
-// acquires the write lock before creating or updating the role at the storage.
-func (b *backend) lockedSetAWSRole(ctx context.Context, s logical.Storage, roleName string, roleEntry *awsRoleEntry) error {
-	if roleName == "" {
-		return fmt.Errorf("missing role name")
+	// Fetch from storage
+	entry, err := s.Get(ctx, "role/"+strings.ToLower(roleName))
+	if err != nil {
+		return nil, err
+	}
+	if entry == nil {
+		return nil, nil
 	}
 
-	if roleEntry == nil {
-		return fmt.Errorf("nil role entry")
+	var result awsRoleEntry
+	if err := entry.DecodeJSON(&result); err != nil {
+		return nil, err
 	}
 
-	b.roleMutex.Lock()
-	defer b.roleMutex.Unlock()
+	needUpgrade, err := b.upgradeRole(ctx, s, &result)
+	if err != nil {
+		return nil, errwrap.Wrapf("error upgrading roleEntry: {{err}}", err)
+	}
+	if needUpgrade && (b.System().LocalMount() || !b.System().ReplicationState().HasState(consts.ReplicationPerformanceSecondary|consts.ReplicationPerformanceStandby)) {
+		if err = b.setRole(ctx, s, roleName, &result); err != nil {
+			return nil, errwrap.Wrapf("error saving upgraded roleEntry: {{err}}", err)
+		}
+	}
 
-	return b.nonLockedSetAWSRole(ctx, s, roleName, roleEntry)
+	return &result, nil
```

Comment:
Do we not want to insert the role in cache after reading from storage?

Reply:
Ah yes! Will fix.